src/compiler/checker.ts

```diff
@@ -6785,30 +6830,175 @@ module ts {
+                    Debug.fail("Property decorators should not have a third synthetic argument.");
+                    break;
+
+                case SyntaxKind.MethodDeclaration:
+                case SyntaxKind.GetAccessor:
+                case SyntaxKind.SetAccessor:
+                    // The `descriptor` for a method decorator will be a `TypedPropertyDescriptor<T>`
+                    // for the type of the member.
+                    let propertyType: Type = getTypeOfNode(parent);
+                    return createTypedPropertyDescriptorType(propertyType);
+            }
+        }
+        
+        function checkApplicableSignature(node: CallLikeExpression, args: Expression[], signature: Signature, relation: Map<RelationComparisonResult>, excludeArgument: boolean[], reportErrors: boolean, containingMessageChain?: DiagnosticMessageChain) {
             for (let i = 0; i < args.length; i++) {
                 let arg = args[i];
                 if (arg.kind !== SyntaxKind.OmittedExpression) {
                     // Check spread elements against rest type (from arity check we know spread argument corresponds to a rest parameter)
                     let paramType = getTypeAtPosition(signature, i);
-                    // A tagged template expression provides a special first argument, and string literals get string literal types
+
+                    // Decorators provide special arguments, a tagged template expression provides 
+                    // a special first argument, and string literals get string literal types
                     // unless we're reporting errors
-                    let argType = i === 0 && node.kind === SyntaxKind.TaggedTemplateExpression
-                        ? globalTemplateStringsArrayType
-                        : arg.kind === SyntaxKind.StringLiteral && !reportErrors
-                            ? getStringLiteralType(<StringLiteral>arg)
-                            : checkExpressionWithContextualType(arg, paramType, excludeArgument && excludeArgument[i] ? identityMapper : undefined);
+                    let argType: Type;
+                    if (arg.parent.kind === SyntaxKind.Decorator) {
+                        if (i === 0) {
+                            argType = getDecoratorTargetArgumentType(arg.parent.parent);
+                        }
+                        else if (i === 1) {
+                            argType = getDecoratorPropertyKeyArgumentType(arg.parent.parent);
+                        }
+                        else if (i === 2) {
+                            argType = getDecoratorPropertyIndexOrDescriptorArgumentType(arg.parent.parent);
+                        }
+                    }
+                    else if (i === 0 && arg.parent.kind === SyntaxKind.TaggedTemplateExpression) {
+                        argType = globalTemplateStringsArrayType;
+                    }
+
+                    if (argType === undefined) {

```

Comment:
I guess that's true, because you'd have to return something that allows to distinguish not having a decorator/template expression versus getting nothing back when you call the argumentType helpers.

Using undefined is okay, provided that decorators and template expressions can *never* result in undefined.